### PR TITLE
[Model] tool calling support for ibm-granite/granite-20b-functioncalling

### PR DESCRIPTION
**Pull request by wseaton**
Add tool calling parser support for `ibm-granite/granite-20b-functioncalling`. 

Also adds an example chat template that is based off of the [granite fc paper](https://arxiv.org/abs/2407.00121).

### CODE REVIEW

File: examples/tool_chat_template_granite.jinja
Comment (wseaton) [reply examples/tool_chat_template_granite.jinja:35]: Good callout! In my evals, I have mostly doing zero-shot single turn calls. 

Let me update the example template with the more general conversational one since that is probably what most people want.

**Comment by njhill** [comment]
@wseaton one of the tests is failing and looks most likely related: https://buildkite.com/vllm/fastcheck/builds/3985#0191dd9c-724a-4fbe-99d3-a1e1c94a2106

**Comment by wseaton** [comment]
@maxdebayser are you still interested in contributing the streaming JSON parser for granite support? I have rebased off of main and might have some bandwidth to work on it, just let me know :)

**Comment by maxdebayser** [comment]
Hi @wseaton , yes, I've been working on this. There is only one unit test that isn't passing. But I know what the cause is.

**Comment by maxdebayser** [comment]
@wseaton, now all tests are passing. I've pushed a squashed merge of everything to this draft PR so that you can take a look and perhaps merge with yours: https://github.com/vllm-project/vllm/pull/8915

**Comment by wseaton** [comment]
@maxdebayser I've cherry picked the commits from your PR, please review!

File: tests/tool_use/utils.py
Comment (maxdebayser) [inline tests/tool_use/utils.py:83]: This one here is a duplicate.
```suggestion
```

**Review by maxdebayser — APPROVED** [review]
Thanks a lot @wseaton . There is an extra entry on the unit test model dict, but other than that it looks good to me.

cc @njhill

**Comment by wseaton** [comment]
@maxdebayser working on fixing merge conflicts w/ upstream now

File: vllm/entrypoints/openai/tool_parsers/granite_20b_fc_tool_parser.py
Comment (maxdebayser) [inline vllm/entrypoints/openai/tool_parsers/granite_20b_fc_tool_parser.py:26]: With the recent changes in main this annotation must be added

```suggestion
@ToolParserManager.register_module("granite-20b-fc")
```

File: vllm/entrypoints/openai/tool_parsers/granite_20b_fc_tool_parser.py
Comment (maxdebayser) [inline vllm/entrypoints/openai/tool_parsers/granite_20b_fc_tool_parser.py:15]:
```suggestion
    ToolParser, ToolParserManager)
```

**Comment by maxdebayser** [comment]
@K-Mistele , if you have the time, could you take a look at this PR?

File: vllm/entrypoints/openai/tool_parsers/utils.py
Comment (njhill) [inline vllm/entrypoints/openai/tool_parsers/utils.py:98]: Could you add typing to these functions? (sorry I know they were just moved in this PR!)

File: vllm/entrypoints/openai/tool_parsers/utils.py
Comment (njhill) [inline vllm/entrypoints/openai/tool_parsers/utils.py:105]: nit: redundant else

**Review by njhill — APPROVED** [review]
Thanks alot @wseaton and @maxdebayser for the reviews

**Comment by njhill** [comment]
@K-Mistele would be great if you could give this a quick look over and stamp :pray:

**Comment by wseaton** [comment]
Getting an OOM error on the granite20b CI checks, looks like the model doesn't quite fit into the available memory of the instance type used for that CI.

File: docs/source/serving/openai_compatible_server.md
Comment (njhill) [inline docs/source/serving/openai_compatible_server.md:256]: @wseaton should `--chat-template` be added here? (like other examples above)

**Comment by njhill** [comment]
> Getting an OOM error on the granite20b CI checks, looks like the model doesn't quite fit into the available memory of the instance type used for that CI.

@wseaton we may need to comment this test out for now, as long as it's been verified manually on a larger GPU.

**Comment by maxdebayser** [comment]
The test could also be run on a larger GPU, although that's probably a bit wasteful
```
$ git diff .
diff --git a/.buildkite/test-pipeline.yaml b/.buildkite/test-pipeline.yaml
index 8c98aa36a..8a396fada 100644
--- a/.buildkite/test-pipeline.yaml
+++ b/.buildkite/test-pipeline.yaml
@@ -293,6 +293,7 @@ steps:
 - label: OpenAI-Compatible Tool Use # 20 min
   fast_check: false
   mirror_hardwares: [ amd ]
+  gpu: a100
   source_file_dependencies:
     - vllm/
     - tests/tool_use
```

File: docs/source/serving/openai_compatible_server.md
Comment (wseaton) [reply docs/source/serving/openai_compatible_server.md:256]: Yes, looks like I dropped it via merge, sorry about that

**Comment by wseaton** [comment]
@njhill Happy to rebuild my fork of vLLM and re-verify the PR manually on an A100, disabled the test for now

**Comment by maxdebayser** [comment]
@wseaton I've verified a couple of hours ago. The tool_use tests run successfully on an A100. I have a quantized version of the model that passes the tests and I'm testing it if it can fit in the CI environment.

**Comment by njhill** [comment]
> The test could also be run on a larger GPU, although that's probably a bit wasteful
> 

@maxdebayser yes that's what I thought too, since none of the other model sub-tests require it

**Comment by njhill** [comment]
Thanks @wseaton! Could you try merging in the latest main branch again? That should hopefully help with the failing CI tests.

**Comment by K-Mistele** [comment]
@wseaton just a minor point of clarification, this PR is for Granite 20B functioncalling -- is this the same function calling format as the granite 3.0 models? I saw that these were released recently:

- https://huggingface.co/collections/ibm-granite/granite-30-models-66fdb59bbb54785c3512114f
- https://huggingface.co/ibm-granite/granite-3.0-8b-instruct

I'm not as familiar with IBM/granite as with some other families of models. if they are compatible, then that's great and we should indicate that in the docs, as well as name the tool parser `granite` most likely. If not (and this is what I suspect is the case, based on a brief reading of their docs & github code for agentic stuff), then it might be worth adding a call-out in the docs that this is NOT compatible with granite 3.0

**Comment by maxdebayser** [comment]
@K-Mistele , they are not compatible. We have another [PR](https://github.com/vllm-project/vllm/pull/9027) for those models but we're waiting for this one to be merged first because there are some code dependencies.

**Comment by maxdebayser** [comment]
@njhill , @wseaton my tests with a quantized version of the model are now [passing](https://github.com/vllm-project/vllm/pull/9700). With quantization the model size was reduced to ~20GB: `mbayser/granite-20b-functioncalling-FP8-KV`, but I still had to add a bunch of flags to reduce memory usage:
`"--max_num_seqs", "1", "--max-model-len", "1024", "--enforce-eager", "--cpu-offload-gb", "20"`

**Comment by wseaton** [comment]
Because of the nature of this branch and how long it's been running (with many merge commits from  main), I don't feel comfortable doing a rebase to fix the DCO issue. Can it be bypassed to get this merged?

**Comment by K-Mistele** [comment]
> @K-Mistele , they are not compatible. We have another [PR](https://github.com/vllm-project/vllm/pull/9027) for those models but we're waiting for this one to be merged first because there are some code dependencies.

sounds good! Let me know if you'd like for me to take a look at it  whenever it's ready :)

**Review by njhill — APPROVED** [review]
Thanks @wseaton @maxdebayser for all of the work on this and thanks @K-Mistele for also reviewing.

**Comment by njhill** [comment]
@wseaton I'll merge this to unblock the other granite PR, could consider re-enabling the test with @maxdebayser's quantized model as a follow-on update...

**Pull request merged**
